### ISSUE TITLE
torch.no_grad() yields NaN values on mps device, 4D attention mask

### ISSUE THREAD

**Issue by AmitMY** [pytorch/pytorch#167515]
### 🐛 Describe the bug

on macOS, define a 4D attention mask that is padded, and call a model.
```py
@pytest.mark.parametrize("device", ["cpu", "mps"])
def test_isolated_language_model_mps(device):
    model = AutoModelForCausalLM.from_pretrained("sbintuitions/tiny-lm").to(device)
    input_ids = torch.tensor([[0, 1, 0, 0], [0, 1, 2, 3]], device=device)

    attention_mask = torch.tensor([
        [[[ True, False, False, False],
          [ True,  True, False, False],
          [False, False, False, False],
          [False, False, False, False]]],
        [[[ True, False, False, False],
          [ True,  True, False, False],
          [ True,  True,  True, False],
          [ True,  True,  True,  True]]]], device=device)
    # attention_mask = torch.tensor([[1, 1, 0, 0], [1, 1, 1, 1]], dtype=torch.bool, device=device)

    with torch.no_grad():
       outputs = model(input_ids=input_ids, attention_mask=attention_mask)
    assert not torch.isnan(outputs.logits).any(), "Logits contain NaN values"
```

### Expected behavior

not NaN - same as it is on CPU.

Also, if removing `torch.no_grad()` it is also not NaN.

<img width="291" height="75" alt="Image" src="https://github.com/user-attachments/assets/9d99ff28-c3bd-4ca8-8515-29ba4434d9b2" />

### Versions

Collecting environment information...
PyTorch version: 2.8.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.0.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.3.19.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:54:21) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-26.0.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M4 Pro

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] torch==2.8.0
[pip3] torchvision==0.23.0
[pip3] vit-pytorch==1.11.7
[conda] numpy                        2.2.6            pypi_0              pypi
[conda] torch                        2.8.0            pypi_0              pypi
[conda] torchvision                  0.23.0           pypi_0              pypi
[conda] vit-pytorch                  1.11.7           pypi_0              pypi

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen

**Comment by malfet** [comment]
I can reproduce it with 2.8, but seems to be fixed with 2.9...

**Comment by jhavukainen** [comment]
Looks like it's related to https://github.com/pytorch/pytorch/issues/156707 where some positions are fully masked.

I think the workaround landed in 2.9.0 so the timeline would make sense: https://github.com/pytorch/pytorch/pull/157727
